src/Storages/StorageAggregatingMemory.cpp

```diff
@@ -0,0 +1,513 @@
+
+    ColumnNumbers keys;
+    for (const auto & key : query_analyzer->aggregationKeys())
+        keys.push_back(header_before_aggregation.getPositionByName(key.name));
+
+    AggregateDescriptions aggregates = query_analyzer->aggregates();
+    for (auto & descr : aggregates)
+        if (descr.arguments.empty())
+            for (const auto & name : descr.argument_names)
+                descr.arguments.push_back(header_before_aggregation.getPositionByName(name));
+
+    Aggregator::Params params(header_before_aggregation, keys, aggregates,
+                              false, settings.max_rows_to_group_by, settings.group_by_overflow_mode,
+                              settings.group_by_two_level_threshold,
+                              settings.group_by_two_level_threshold_bytes,
+                              settings.max_bytes_before_external_group_by,
+                              settings.empty_result_for_aggregation_by_empty_set,
+                              select_context->getTemporaryVolume(),
+                              settings.max_threads,
+                              settings.min_free_disk_space_for_temporary_data,
+                              true);
+
+    aggregator_transform = std::make_shared<AggregatingTransformParams>(params, false);
+
+    initState(constructor_context);
+
+    is_initialized = true;
+}
+
+void StorageAggregatingMemory::initState(ContextPtr context)
+{
+    many_data = std::make_shared<ManyAggregatedData>(1);
+
+    /// If there was no data, and we aggregate without keys, and we must return single row with the result of empty aggregation.
+    /// To do this, we pass a block with zero rows to aggregate.
+    if (aggregator_transform->params.keys_size == 0 && !aggregator_transform->params.empty_result_for_aggregation_by_empty_set)
+    {
+        AggregatingOutputStream os(*this, getInMemoryMetadataPtr(), context);
+        os.write(src_block_header);
+    }
+}
+
+void StorageAggregatingMemory::startup()
+{
+    try
+    {
+        lazyInit();
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("Failed to initialize AggregatingMemory on startup");
+        LOG_ERROR(log, "{}", getCurrentExceptionMessage(true));
+    }
+}
+
+Pipe StorageAggregatingMemory::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t /*max_block_size*/,
+    unsigned num_streams)
+{
+    lazyInit();
+    metadata_snapshot->check(column_names, getVirtuals(), getStorageID());
+    // TODO implement O(1) read by aggregation key
+
+    auto prepared_data = aggregator_transform->aggregator.prepareVariantsToMerge(many_data->variants);
+    auto prepared_data_ptr = std::make_shared<ManyAggregatedDataVariants>(std::move(prepared_data));
+
+    ProcessorPtr source;
+
+    auto filter_key = getFilterKeys(aggregator_transform->params, query_info);
```

Comment:
Looks like it is not used anymore.
Also, I don't see where `Aggregator::readBlockByFilterKeys` is used.